src/model_theory/terms_and_formulas.lean

```diff
@@ -34,300 +34,333 @@ the continuum hypothesis*][flypitch_itp]
+  (⊤ : L.bounded_formula α l).realize v xs ↔ true :=
 by simp [has_top.top]
 
-@[simp] lemma realize_inf : realize_bounded_formula M (φ ⊓ ψ) v xs =
-    (realize_bounded_formula M φ v xs ∧ realize_bounded_formula M ψ v xs) :=
-by simp [has_inf.inf, realize_bounded_formula]
+@[simp] lemma realize_inf : (φ ⊓ ψ).realize v xs ↔ (φ.realize v xs ∧ ψ.realize v xs) :=
+by simp [has_inf.inf, realize]
 
-@[simp] lemma realize_bd_imp : realize_bounded_formula M (φ.bd_imp ψ) v xs =
-    (realize_bounded_formula M φ v xs → realize_bounded_formula M ψ v xs) :=
-by simp only [realize_bounded_formula]
+@[simp] lemma realize_imp : (φ.imp ψ).realize v xs ↔ (φ.realize v xs → ψ.realize v xs) :=
+by simp only [realize]
 
-@[simp] lemma realize_bd_rel {k : ℕ} (R : L.relations k) (ts : fin k → L.term _) :
-  realize_bounded_formula M (bd_rel R ts) v xs =
-    rel_map R (λ i, realize_term (sum.elim v xs) (ts i)) :=
-rfl
+@[simp] lemma realize_rel {k : ℕ} {R : L.relations k} {ts : fin k → L.term _} :
+  (R.bounded_formula ts).realize v xs ↔ rel_map R (λ i, (ts i).realize (sum.elim v xs)) :=
+iff.refl _
 
-@[simp] lemma realize_sup : realize_bounded_formula M (φ ⊔ ψ) v xs =
-    (realize_bounded_formula M φ v xs ∨ realize_bounded_formula M ψ v xs) :=
+@[simp] lemma realize_sup : (φ ⊔ ψ).realize v xs ↔ (φ.realize v xs ∨ ψ.realize v xs) :=
 begin
-  simp only [realize_bounded_formula, has_sup.sup, realize_bd_not, eq_iff_iff],
+  simp only [realize, has_sup.sup, realize_not, eq_iff_iff],
   tauto,
 end
 
-@[simp] lemma realize_bd_all : realize_bounded_formula M (bd_all θ) v xs =
-    ∀ (a : M), (realize_bounded_formula M θ v (fin.snoc xs a)) :=
-rfl
+@[simp] lemma realize_all : (all θ).realize v xs ↔ ∀ (a : M), (θ.realize v (fin.snoc xs a)) :=
+iff.refl _
 
-@[simp] lemma realize_bd_exists : realize_bounded_formula M θ.bd_exists v xs =
-    ∃ (a : M), (realize_bounded_formula M θ v (fin.snoc xs a)) :=
+@[simp] lemma realize_ex : θ.ex.realize v xs ↔ ∃ (a : M), (θ.realize v (fin.snoc xs a)) :=
 begin
-  rw [bd_exists, realize_bd_not, realize_bd_all, not_forall],
-  simp_rw [realize_bd_not, not_not],
+  rw [bounded_formula.ex, realize_not, realize_all, not_forall],
+  simp_rw [realize_not, not_not],
 end
 
-@[simp] lemma realize_bd_iff : realize_bounded_formula M (φ.bd_iff ψ) v xs =
-  (realize_bounded_formula M φ v xs ↔ realize_bounded_formula M ψ v xs) :=
+@[simp] lemma realize_iff : (φ.iff ψ).realize v xs ↔ (φ.realize v xs ↔ ψ.realize v xs) :=
+by simp only [bounded_formula.iff, realize_inf, realize_imp, and_imp, ← iff_def]
+
+lemma realize_relabel {m n : ℕ}
+  {φ : L.bounded_formula α n} {g : α → (β ⊕ fin m)} {v : β → M} {xs : fin (m + n) → M} :
+  (φ.relabel g).realize v xs ↔
+    φ.realize (sum.elim v (xs ∘ (fin.cast_add n)) ∘ g) (xs ∘ (fin.nat_add m)) :=
 begin
-  rw [bd_iff, iff_def],
-  simp,
+  have h : ∀ (n : ℕ) (xs : fin (m + n) → M), sum.elim v xs ∘ bounded_formula.relabel_aux g n =
+    sum.elim (sum.elim v (xs ∘ (fin.cast_add n)) ∘ g) (xs ∘ (fin.nat_add m)),
```

Comment:
I think it's pretty likely that you will need lemmas like this about `relabel_aux` more frequently. Maybe it's nicer to separate them out into actual lemmas. Doesn't have to be done in this PR.